src/main/java/uk/gov/companieshouse/itemhandler/kafka/OrdersKafkaConsumer.java

```diff
@@ -1,173 +1,271 @@
+     * method to process received message. If the `retryable` processor is unsuccessful with a `retryable` error, after
+     * maximum numbers of attempts allowed, the message is republished to `-retry` topic for failover processing.
+     * This listener stops accepting messages when the topic's offset reaches `errorRecoveryOffset`.
+     * @param message
+     */
+    @KafkaListener(id = ORDER_RECEIVED_GROUP_ERROR, groupId = ORDER_RECEIVED_GROUP_ERROR,
+                    topics = ORDER_RECEIVED_TOPIC_ERROR,
+                    autoStartup = "${uk.gov.companieshouse.item-handler.error-consumer}")
+    public void processOrderReceivedError(org.springframework.messaging.Message<OrderReceived> message) {
+        long offset = Long.parseLong("" + message.getHeaders().get("kafka_offset"));
+        if (offset <= errorRecoveryOffset) {
+            handleMessage(message);
+        }
+        else {
+            LOGGER.info(String.format("Pausing error consumer \"%1$s\" as error recovery offset '%2$d' reached.",
+                    ORDER_RECEIVED_GROUP_ERROR, errorRecoveryOffset));
+            registry.getListenerContainer(ORDER_RECEIVED_GROUP_ERROR).pause();
         }
     }
 
-    @KafkaListener(topics = ORDER_RECEIVED_TOPIC_RETRY, groupId = ORDER_RECEIVED_GROUP_RETRY,
-            autoStartup = "#{!${uk.gov.companieshouse.item-handler.error-consumer}}")
-    public void processOrderReceivedRetry(String orderReceivedUri)
-            throws SerializationException, ExecutionException, InterruptedException {
+    /**
+     * Handles processing of received message.
+     * @param message
+     */
+    protected void handleMessage(org.springframework.messaging.Message<OrderReceived> message) {
+        OrderReceived msg = message.getPayload();
+        String orderReceivedUri = msg.getOrderUri();
+        MessageHeaders headers = message.getHeaders();
+        String receivedTopic = headers.get(KafkaHeaders.RECEIVED_TOPIC).toString();
         try {
-            logMessageReceived(orderReceivedUri, ORDER_RECEIVED_TOPIC_RETRY);
-        } catch (Exception ex){
-            republishMessageToTopic(orderReceivedUri, ORDER_RECEIVED_TOPIC_RETRY, ORDER_RECEIVED_TOPIC_ERROR, ex);
+            logMessageReceived(message);
+
+            // process message
+
+            logMessageProcessed(message);
+        } catch (RetryableErrorException ex){
+            String nextTopic = (receivedTopic.equals(ORDER_RECEIVED_TOPIC)
+                    || receivedTopic.equals(ORDER_RECEIVED_TOPIC_ERROR))
+                    ? ORDER_RECEIVED_TOPIC_RETRY : ORDER_RECEIVED_TOPIC_ERROR;
+            String counterKey = receivedTopic + "-" + orderReceivedUri;
+
+            if (receivedTopic.equals(ORDER_RECEIVED_TOPIC)
+                    || retryCount.getOrDefault(counterKey, 0) >= MAX_RETRY_ATTEPTS) {
+                if (!receivedTopic.equals(ORDER_RECEIVED_TOPIC)) {
```

Comment:
I think we can work out that this means clean up the retry count once we have exhausted the allowed number of retry attempts when retrying in the `ORDER_RECEIVED_TOPIC_RETRY` topic? However I have to think quite hard about this and I do wonder whether what is going on could be made clearer with named functions etc?

Reply:
Both `-retry` and `-error` consumers are allowed `MAX_RETRY_ATTEMPTS`, so using the topic name in the `retryKey` helps manage attempts for each of them, individually.

Reply:
I see. OK, so would you say then that according to your understanding the `resilience non error consumer flow` diagram in https://companieshouse.atlassian.net/wiki/spaces/DEV/pages/129564747/Kafka+resilience#Kafkaresilience-HighLevelOverview could do with a tweak?

Reply:
Yes. The diagram definitely needs simplifying.

Reply:
So the pre-existing resilient consumer implementations basically attempt `MAX_RETRY_ATTEMPTS` retries when handling messages from either the `retry` or the `error` topic? They do not attempt any retries for the `main` topic? Instead they shunt a recoverably failing message straight away from the `main` to the `retry` topic?

Reply:
Yes, that is what I have understood from Joanne and the resilience handler diagram.